### ISSUE TITLE
configured values sometimes ignored

### ISSUE THREAD

**Issue by jackd** [google/gin-config#16]
I've been witnessing some inconsistent behaviour with basic configuration - some configs are ignored completely. I wish I could be more specific as to when they are ignored and when they aren't, but I'm at a loss as to any kind of pattern.

Minimal example [here](https://gist.github.com/jackd/f345b1dcd75e0efa53ba912d169852ec)

**Comment by sguada** [comment]
The issue in your example is that you are using an `external_configurable` which doesn't change the behavior of the Class being configured in the code, only when is passed as an argument to another configurable.

In general, changing the behavior of a Class globally can be dangerous, I will recommend to be explicit about what is being configured, for example:

```
gin.config.external_configurable(tf.keras.layers.Dense, module='layers')
gin.config.external_configurable(tf.keras.regularizers.l2, module='regularizers')

@gin.configurable
def my_model(dense_layer):
  x = tf.zeros((5, 4))
  layer = dense_layer(10)
  layer(x)

config = """
layers.Dense.kernel_regularizer = @regularizers.l2()
regularizers.l2.l = 1e-4
my_model.dense_layer = @layers.Dense
"""

gin.parse_config(config)

**Comment by jackd** [comment]
I see - thought it might be a bit risky, but figured it was one of those "community of consenting adults" things. Bit bummed I spent a week looking for patterns in noise after effectively training the same model 40 times, but I've just found the note in the docs explaining this so I guess I only have myself to blame.

For others viewing this with the same issue: to achieve what I was going for, I slightly prefer
```
Dense = gin.config.external_configurable(tf.keras.layers.Dense, module='layers')
layers = Dense(10)  # will have any configured regularizer etc.
```
